File: packages/ios-sdk/Sources/GateFlowKit/Identity/IdentityManager.swift
import Foundation
import Combine

/// Manages user identity: anonymous alias, logged-in userId, and user attributes.
/// Thread-safe via a dedicated dispatch queue.
final class IdentityManager {
    /// The effective userId: appUserId if logged in, otherwise aliasId.
    var userId: String {
        queue.sync { _appUserId ?? _aliasId }
    }

    var appUserId: String? { queue.sync { _appUserId } }
    var aliasId: String { queue.sync { _aliasId } }
    var seed: Int { queue.sync { _seed } }
    var isLoggedIn: Bool { queue.sync { _appUserId != nil } }

    var userAttributes: [String: AnyCodable] {
        queue.sync { _userAttributes }
    }

    /// Publisher that emits `true` when identity is ready.
    let hasIdentity = CurrentValueSubject<Bool, Never>(false)

    private var _appUserId: String? {
        didSet { saveIds() }
    }
    private var _aliasId: String {
        didSet { saveIds() }
    }
    private var _seed: Int {
        didSet { saveIds() }
    }
    private var _userAttributes: [String: AnyCodable] = [:]

    private let queue = DispatchQueue(label: "com.gateflow.identitymanager")
    private let group = DispatchGroup()
    private let storage: Storage
    private let notifyUserChange: (([String: AnyCodable]) -> Void)?

    init(storage: Storage, notifyUserChange: (([String: AnyCodable]) -> Void)? = nil) {
        self.storage = storage
        self.notifyUserChange = notifyUserChange

        self._appUserId = storage.readString(StorageKeys.appUserId)
        self._aliasId = storage.readString(StorageKeys.aliasId) ?? {
            let alias = IdentityLogic.generateAlias()
            storage.writeString(StorageKeys.aliasId, value: alias)
            return alias
        }()
        self._seed = storage.readInt(StorageKeys.seed) ?? {
            let seed = IdentityLogic.generateSeed()
            storage.writeInt(StorageKeys.seed, value: seed)
            return seed
        }()
        if let attrs = storage.readCodable(StorageKeys.userAttributes) as [String: AnyCodable]? {
            self._userAttributes = attrs
        }

        var extraAttributes: [String: AnyCodable] = [:]
        extraAttributes["aliasId"] = AnyCodable(aliasId)
        extraAttributes["seed"] = AnyCodable(seed)
        if let appUserId = _appUserId {
            extraAttributes["appUserId"] = AnyCodable(appUserId)
        }
        if !extraAttributes.isEmpty {
            mergeUserAttributes(extraAttributes, shouldTrackMerge: false)
        }
    }

    func configure(isFirstAppOpen: Bool) async {
        group.enter()
        group.leave()
        didSetIdentity()
    }

    /// Identify the user with a server-side userId.
    func identify(userId: String) {
        guard let sanitized = IdentityLogic.sanitize(userId: userId) else {
            GFLogger.debug(level: .error, scope: .identityManager, message: "Provided userId was empty.")
            return
        }

        group.enter()
        queue.async { [weak self] in
            guard let self = self else { return }

            if self._appUserId == sanitized {
                self.group.leave()
                return
            }

            self.hasIdentity.send(false)

            // If switching users, reset.
            if self._appUserId != nil, sanitized != self._appUserId {
                self._reset()
            }

            self._appUserId = sanitized
            self.group.leave()
            self.didSetIdentity()
        }
    }

    /// Reset to anonymous identity.
    func reset() {
        hasIdentity.send(false)
        group.enter()
        queue.async { [weak self] in
            self?._reset()
            self?.group.leave()
            self?.didSetIdentity()
        }
    }

    func mergeUserAttributes(_ newAttrs: [String: AnyCodable], shouldTrackMerge: Bool = true) {
        queue.async { [weak self] in
            self?._mergeUserAttributes(newAttrs, shouldTrackMerge: shouldTrackMerge)
        }
    }

    private func _reset() {
        _appUserId = nil
        _aliasId = IdentityLogic.generateAlias()
        _seed = IdentityLogic.generateSeed()
        _userAttributes = [:]
    }

    private func didSetIdentity() {
        group.notify(queue: .main) { [weak self] in
            self?.hasIdentity.send(true)
        }
    }

    private func saveIds() {
        if let appUserId = _appUserId {
            storage.writeString(StorageKeys.appUserId, value: appUserId)
        }
        storage.writeString(StorageKeys.aliasId, value: _aliasId)
        storage.writeInt(StorageKeys.seed, value: _seed)

        var newAttrs: [String: AnyCodable] = [
            "aliasId": AnyCodable(_aliasId),
            "seed": AnyCodable(_seed),
        ]
        if let appUserId = _appUserId {
            newAttrs["appUserId"] = AnyCodable(appUserId)
        }
        _mergeUserAttributes(newAttrs, shouldTrackMerge: false)
    }

    private func _mergeUserAttributes(_ newAttrs: [String: AnyCodable], shouldTrackMerge: Bool) {
        var merged = _userAttributes
        for (key, value) in newAttrs {
            merged[key] = value
        }
        _userAttributes = merged
        storage.writeCodable(StorageKeys.userAttributes, value: merged)

        if shouldTrackMerge {
            notifyUserChange?(merged)
        }
    }
}

File: packages/android-sdk/gateflow/src/main/java/com/gateflow/sdk/identity/IdentityManager.kt
package com.gateflow.sdk.identity

import com.gateflow.sdk.experiment.UserAttributes
import com.gateflow.sdk.logger.LogLevel
import com.gateflow.sdk.logger.LogScope
import com.gateflow.sdk.logger.Logger
import com.gateflow.sdk.storage.LocalStorage
import com.gateflow.sdk.storage.StorageKeys
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.filter

/** Manages user identity: anonymous alias, logged-in userId, and user attributes. */
class IdentityManager(
    private val storage: LocalStorage,
) {
    data class IdentityState(
        val appUserId: String? = null,
        val aliasId: String,
        val seed: Int,
        val userAttributes: Map<String, Any?> = emptyMap(),
        val isReady: Boolean = false,
    ) {
        val userId: String get() = appUserId ?: aliasId
        val isLoggedIn: Boolean get() = appUserId != null
        val enrichedAttributes: Map<String, Any?>
            get() = userAttributes + mapOf(
                "aliasId" to aliasId,
                "seed" to seed,
                "appUserId" to appUserId,
            )
    }

    private val _state = MutableStateFlow(
        IdentityState(
            appUserId = storage.read(StorageKeys.AppUserId),
            aliasId = storage.read(StorageKeys.AliasId) ?: generateAndSaveAlias(),
            seed = storage.read(StorageKeys.Seed) ?: generateAndSaveSeed(),
            userAttributes = loadAttributes(),
            isReady = true,
        )
    )

    val state: StateFlow<IdentityState> = _state.asStateFlow()

    val hasIdentity = _state.map { it.isReady }.filter { it }

    fun configure(isFirstAppOpen: Boolean) {
        // For now, identity is ready immediately.
        // In a fuller implementation this would trigger assignment fetching.
        _state.value = _state.value.copy(isReady = true)
    }

    fun identify(userId: String) {
        val sanitized = IdentityLogic.sanitize(userId) ?: run {
            Logger.debug(
                level = LogLevel.ERROR,
                scope = LogScope.IdentityManager,
                message = "Provided userId was empty.",
            )
            return
        }

        val current = _state.value
        if (current.appUserId == sanitized) return

        // If switching users, reset.
        if (current.appUserId != null) {
            reset()
        }

        storage.write(StorageKeys.AppUserId, sanitized)
        _state.value = current.copy(appUserId = sanitized, isReady = false)

        // Mark ready after identity resolution
        _state.value = _state.value.copy(isReady = true)
    }

    fun reset() {
        val current = _state.value
        storage.delete(StorageKeys.AppUserId)
        val newAlias = generateAndSaveAlias()
        val newSeed = generateAndSaveSeed()
        storage.delete(StorageKeys.UserAttributes)

        _state.value = IdentityState(
            appUserId = null,
            aliasId = newAlias,
            seed = newSeed,
            isReady = true,
        )
    }

    fun mergeUserAttributes(newAttrs: Map<String, Any?>, shouldTrackMerge: Boolean = false) {
        val current = _state.value
        val merged = current.userAttributes + newAttrs
        val json = serializeAttributes(merged)
        storage.write(StorageKeys.UserAttributes, json)
        _state.value = current.copy(userAttributes = merged)

        if (shouldTrackMerge) {
            Logger.debug(
                level = LogLevel.INFO,
                scope = LogScope.IdentityManager,
                message = "User attributes updated",
                info = mapOf("count" to merged.size),
            )
        }
    }

    private fun generateAndSaveAlias(): String {
        val alias = IdentityLogic.generateAlias()
        storage.write(StorageKeys.AliasId, alias)
        return alias
    }

    private fun generateAndSaveSeed(): Int {
        val seed = IdentityLogic.generateSeed()
        storage.write(StorageKeys.Seed, seed)
        return seed
    }

    private fun loadAttributes(): Map<String, Any?> {
        val json = storage.read(StorageKeys.UserAttributes) ?: return emptyMap()
        return deserializeAttributes(json)
    }

    // Simple JSON serialization via moshi is handled at the Network layer;
    // here we use a basic approach for primitive maps.
    private fun serializeAttributes(attrs: Map<String, Any?>): String {
        return attrs.entries.joinToString(separator = ",", prefix = "{", postfix = "}") { (k, v) ->
            val vs = when (v) {
                null -> "null"
                is String -> "\"$v\""
                else -> v.toString()
            }
            "\"$k\":$vs"
        }
    }

    private fun deserializeAttributes(json: String): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        val trimmed = json.trimStart('{').trimEnd('}').trim()
        if (trimmed.isEmpty()) return map

        // Very basic parser — sufficient for flat string/int/bool maps
        trimmed.split(",").forEach { pair ->
            val parts = pair.split(":", limit = 2)
            if (parts.size == 2) {
                val key = parts[0].trim('"')
                val raw = parts[1].trim()
                val value = when {
                    raw == "null" -> null
                    raw == "true" -> true
                    raw == "false" -> false
                    raw.toIntOrNull() != null -> raw.toInt()
                    raw.toDoubleOrNull() != null -> raw.toDouble()
                    else -> raw.trim('"')
                }
                map[key] = value
            }
        }
        return map
    }
}
